core/imds/src/test/java/software/amazon/awssdk/imds/Ec2MetadataClientTest.java

```diff
@@ -0,0 +1,391 @@
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).inScenario("Retry Scenario")
+                                                        .whenScenarioStateIs("Cause Success")
+                                                        .willReturn(aResponse().withBody("some-token")));
+
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).inScenario("Retry Scenario")
+                                                    .whenScenarioStateIs("Cause Success")
+                                                    .willReturn(aResponse().withBody("{}")));
+
+
+        MetadataResponse metadataResponse = ec2MetadataClient.get("/latest/meta-data/ami-id");
+        assertThat(metadataResponse.asString()).isEqualTo("{}");
+
+        WireMock.verify(2, putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH))
+            .withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("21600")));
+        WireMock.verify(exactly(1), getRequestedFor(urlPathEqualTo(AMI_ID_RESOURCE))
+            .withHeader(TOKEN_HEADER, equalTo("some-token")));
+    }
+
+    @Test
+    public void getToken_failedTwice_shouldSucceedOnThirdAttempt() {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).inScenario("Retry Scenario")
+                                                        .whenScenarioStateIs(STARTED)
+                                                        .willReturn(aResponse().withFault(MALFORMED_RESPONSE_CHUNK))
+                                                        .willSetStateTo("Try-2"));
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).inScenario("Retry Scenario")
+                                                        .whenScenarioStateIs("Try-2")
+                                                        .willReturn(aResponse().withFault(CONNECTION_RESET_BY_PEER))
+                                                        .willSetStateTo("Try-3"));
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).inScenario("Retry Scenario")
+                                                        .whenScenarioStateIs("Try-3")
+                                                        .willReturn(aResponse().withBody("some-token")));
+
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE))
+                    .withHeader(TOKEN_HEADER, equalTo("some-token"))
+                    .willReturn(aResponse().withBody("{}")));
+
+        MetadataResponse metadataResponse = ec2MetadataClient.get("/latest/meta-data/ami-id");
+        assertThat(metadataResponse.asString()).isEqualTo("{}");
+
+        WireMock.verify(3, putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH))
+            .withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("21600")));
+        WireMock.verify(exactly(1), getRequestedFor(urlPathEqualTo(AMI_ID_RESOURCE))
+            .withHeader(TOKEN_HEADER, equalTo("some-token")));
+    }
+
+    @Test
+    public void getToken_failsTwiceWithIOExceptionThenOK_shouldSucceedOnThirdAttempt() {
```

Comment:
Do all permutations of "fail a couple of times and then succeed" test cases test different behaviors? Could we pare them down a little bit?

Reply:
The idea is to test these cases:
- token request succeed (with or without retry) then request succeed.
- token request succeed (with or without retry) then request fails.
- token request fails (with or without retry) then request succeed. 
- token request fails (with or without retry) then request fails.